.github/workflows/common/common_pytest.yml

```diff
@@ -0,0 +1,39 @@
+# A common CI workflow that runs pytest on all actively supported Python versions.
+#
+# Assumptions:
+#
+# 1. The repo works with the below-specified version of Poetry.
+# 2. The repo can be tested with pytest. 
+# 3. All Python versions named below are enabled for testing.
+
+name: Python Pytest Check (MeltanoLabs)
+
+on: [push]
```

Comment:
```suggestion
on:
  pull_request: {}
  push:
    branches: [main]
  workflow_dispatch:
    inputs: {}

concurrency:
  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
  cancel-in-progress: true
```

Reply:
I don't know if I follow the distinctions here. 

- Workflow dispatches to allow these to be run manually?
- I could guess but I'm not confident I know what the `concurrency` block does.
- What's the benefit of `pull_request: {}, push: { branches: [main] }` vs just `on: [push]`?

Reply:
`on: push` triggers a workflow run for each push, even if there is no associated PR.

The suggested `on` block triggers a workflow run for each update to the PR (e.g. each time new commits are pushed), and will additionally run the workflow when `main` is pushed to. It also support running the workflow manually, as you said.

The `concurrency` block groups running jobs if they have the same `group`, which is `${{ github.workflow }}-${{ github.head_ref || github.run_id }}`, e.g. `common_pytest-1234-branch-name`, or `common_pytest-12341251346132346`. When a new job named ABC is started that matches that group, and an old job ABC within the group is still in-progress, the old one is cancelled. This is so that if someone pushes multiple times to a PR in quick succession, only the most recent jobs will run to completion, since the old ones are obsolete. We use this same block of code in several Meltano repositories.

I'll update it with a comment explaining it.

Reply:
> `on: push` triggers a workflow run for each push, even if there is no associated PR.

I think this was my desired behavior, to run on all commits even if a PR doesn't (yet) exist and including `main` branch along with all other branches. Any reason _not_ to run if a PR doesn't exist yet?

Reply:
FWIW this may trigger duplicate runs once the PR is created, one for the push and another for the PR sync.

Reply:
There isn't much visibility when automatically running workflows without a PR, and it may waste compute (which we don't pay for for public repos, but it still feels wasteful).

I have no strong objection to using `on: push` (plus `workflow_dispatch`) if you think running on every push is valuable.

Reply:
> This is so that if someone pushes multiple times to a PR in quick succession, only the most recent jobs will run to completion, since the old ones are obsolete. We use this same block of code in several Meltano repositories.

Okay, this makes sense. I don't personally mind those jobs running to completion for taps/targets where the duration is rather short. In the past I have pushed a commit followed by a second one and maybe a third, specifically to see how the variations will perform in CI, essentially parallelizing the work by offloading it.

For these short-running workloads, I'm fine to use the dedupe / short-circuit logic you propose. I slightly prefer not to dedupe but not strongly enough to push back if that's your prefered behavior.